driver-core/src/main/java/com/datastax/driver/core/policies/ErrorAwarePolicy.java

```diff
@@ -0,0 +1,401 @@
+/*
+ *      Copyright (C) 2012-2015 DataStax Inc.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ */
+package com.datastax.driver.core.policies;
+
+import com.datastax.driver.core.Cluster;
+import com.datastax.driver.core.ErrorTracker;
+import com.datastax.driver.core.Host;
+import com.datastax.driver.core.HostDistance;
+import com.datastax.driver.core.Statement;
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.AbstractIterator;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+
+import java.util.ArrayDeque;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Queue;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Chainable load balancing policy that will filter out hosts based on their errors rates.
+ * <p/>
+ * The policy bases its list of hosts on a child policy on which it will compute whether this host needs to included
+ * or excluded of the query plan.
+ * <p/>
+ * The policy uses a {@link ErrorAwarePolicyTracker} which is an extension of the base {@link ErrorTracker} that it
+ * will register automatically on the {@link Cluster}. The {@link ErrorAwarePolicyTracker} will be regularly updated
+ * and the policy will refer to it to take a decision on whether to include the child policy query plan's next host
+ * in the query plan. The policy will schedule a {@link ErrorTrackerUpdater} task on a fixed period that will call
+ * on each invocation {@link ErrorAwarePolicyTracker#update()}.
+ * <p/>
+ * The policy follows the builder pattern to be created, the {@link Builder} class can be created with
+ * {@link #builder} method.
+ */
+public class ErrorAwarePolicy implements ChainableLoadBalancingPolicy {
+
+    private final LoadBalancingPolicy childPolicy;
+
+    private final long updateRate;
+
+    private final static String EXECUTOR_SERVICE_NAME = "ErrorTracker updater task";
+
+    private final ScheduledExecutorService updaterService = Executors.newSingleThreadScheduledExecutor(new ThreadFactoryBuilder().setNameFormat(EXECUTOR_SERVICE_NAME).build());
```

Comment:
It's also used for cleaning up idle connections.   Was thinking that if the activity done within the executor is not intensive / low risk of blocking that we could used a shared executor.  One -1 for that is that the behavior is driven by ErrorAwarePolicyTracker#update, which can be user defined, so maybe best to avoid this possible optimization for now.